hail/src/main/scala/is/hail/expr/ir/EmitStream.scala

```diff
@@ -8,9 +8,399 @@ import is.hail.expr.types.physical._
+        close0 = Code(leftSource.close0, rightSource.close0),
+        setup = Code(leftSource.setup, rightSource.setup),
+        close = Code(leftSource.close, rightSource.close),
+        pull = leftSource.pull)
+    }
+  }
+
+  def multiZip(streams: IndexedSeq[Stream[Code[_]]]): Stream[IndexedSeq[Code[_]]] = new Stream[IndexedSeq[Code[_]]] {
+    def apply(eos: Code[Ctrl], push: IndexedSeq[Code[_]] => Code[Ctrl])(implicit ctx: EmitStreamContext): Source[IndexedSeq[Code[_]]] = {
+      val closeJPs = streams.map(_ => joinPoint())
+      val eosJP = closeJPs(0)
+      val terminatingStream = newLocal[Code[Int]]
+
+      def nthSource(n: Int, acc: IndexedSeq[Code[_]]): (Source[Code[_]], IndexedSeq[Code[Unit]]) = {
+        if (n == streams.length - 1) {
+          val src = streams(n)(
+            Code(terminatingStream := n, eosJP(())),
+            c => push(acc :+ c))
+          (src, IndexedSeq(src.close))
+        } else {
+          var rest: Source[Code[_]] = null
+          var closes: IndexedSeq[Code[Unit]] = null
+          val src = streams(n)(
+            Code(terminatingStream := n, eosJP(())),
+            c => {
+              val t = nthSource(n+1, acc :+ c)
+              rest = t._1
+              closes = t._2
+              rest.pull
+            })
+          (Source[Code[_]](
+              setup0 = Code(src.setup0, rest.setup0),
+              close0 = Code(rest.close0, src.close0),
+              setup = Code(src.setup, rest.setup),
+              close = Code(rest.close, src.close),
+              pull = src.pull),
+            src.close +: closes)
+        }
+      }
+
+      val (source, closes) = nthSource(0, IndexedSeq.empty)
+
+      closeJPs.zipWithIndex.foreach { case (jp, i) =>
+        val next = if (i == streams.length - 1) eos else closeJPs(i + 1)(())
+        jp.define(_ => terminatingStream.load.ceq(i).mux(
+          next,
+          Code(closes(i), next)))
+      }
+
+      source.asInstanceOf[Source[IndexedSeq[Code[_]]]]
+    }
+  }
+
+  def mux[A: ParameterPack](cond: Code[Boolean], left: Stream[A], right: Stream[A]): Stream[A] = new Stream[A] {
+    def apply(eos: Code[Ctrl], push: A => Code[Ctrl])(implicit ctx: EmitStreamContext): Source[A] = {
+      val b = newLocal[Code[Boolean]]
+      val eosJP = joinPoint()
+      val pushJP = joinPoint[A]
+
+      eosJP.define(_ => eos)
+      pushJP.define(push)
+
+      val l = left(eosJP(()), pushJP(_))
+      val r = right(eosJP(()), pushJP(_))
+
+      val lPullJP = joinPoint()
+      val rPullJP = joinPoint()
+
+      lPullJP.define(_ => l.pull)
+      rPullJP.define(_ => r.pull)
+      Source[A](
+        setup0 = Code(b := cond, l.setup0, r.setup0),
+        close0 = Code(l.close0, r.close0),
+        setup = b.load.mux(l.setup, r.setup),
+        close = b.load.mux(l.close, r.close),
+        pull = JoinPoint.mux(b.load, lPullJP, rPullJP)
+      )
+    }
+  }
+
+  def fromParameterized[P, A](
+    stream: EmitStream.Parameterized[P, A]
```

Comment:
Just want to verify my understranding, this will go away when EmitStream2 is complete over and EmitStream(1) is removed, right?

Reply:
Right.